src/astarte-client/transforms/dataTree.ts

```diff
@@ -0,0 +1,346 @@
+        },
+      ];
+    } else if (iface.type === 'datastream' && iface.aggregation === 'individual') {
+      const interfaceData = data as AstarteDatastreamIndividualData[];
+      this.linearizedData = interfaceData.map((obj) => ({
+        endpoint: obj.endpoint,
+        timestamp: obj.timestamp,
+        ...({ type: obj.type, value: obj.value } as AstarteDataTuple),
+      }));
+    } else {
+      const interfaceData = data as AstarteDatastreamObjectData[];
+      this.linearizedData = interfaceData
+        .map((obj) =>
+          Object.entries(obj.value).map(([prop, propValue]) => ({
+            endpoint: `${obj.endpoint}/${prop}`,
+            timestamp: obj.timestamp,
+            ...propValue,
+          })),
+        )
+        .flat();
+    }
+  }
+
+  getNode(endpoint: string): AstarteDataTreeLeaf<Data> | null {
+    const sanitizedEndpoint = endpoint.replace(/\/$/, '');
+    if (sanitizedEndpoint === this.endpoint) {
+      return this;
+    }
+    return null;
+  }
+
+  getLeaves(): AstarteDataTreeLeaf<Data>[] {
+    return [this];
+  }
+}
+
+interface AstarteDataTreeRootParams<
+  Data extends AstartePropertyData | AstarteDatastreamIndividualData | AstarteDatastreamObjectData,
+  Interface extends AstarteInterface
+> {
+  interface: Interface;
+  data: AstarteInterfaceValues;
+  endpoint?: string;
+  relativeEndpoint?: string;
+  parentNode?: AstarteDataTreeRoot<Data> | null;
+}
+class AstarteDataTreeRoot<
+  Data extends AstartePropertyData | AstarteDatastreamIndividualData | AstarteDatastreamObjectData,
+  Interface extends AstarteInterface = AstarteInterface
+> {
+  readonly parent: AstarteDataTreeRoot<Data> | null;
+
+  readonly children: Array<AstarteDataTreeRoot<Data> | AstarteDataTreeLeaf<Data>>;
+
+  readonly endpoint: string;
+
+  readonly relativeEndpoint: string;
+
+  readonly dataKind: AstarteDataTreeKind;
+
+  readonly rawData: AstarteInterfaceValues;
+
+  constructor({
+    interface: iface,
+    data,
+    endpoint = '',
+    relativeEndpoint = '',
+    parentNode = null,
+  }: AstarteDataTreeRootParams<Data, Interface>) {
+    this.endpoint = endpoint;
+    this.relativeEndpoint = relativeEndpoint;
+    this.parent = parentNode;
+    this.dataKind = getDataTreeKind(iface);
+    this.rawData = data;
+    this.children = Object.entries(data)
+      .map(([prop, propValue]) => {
+        if (iface.type === 'properties' && isPropertiesInterfaceValue(propValue)) {
+          return new AstarteDataTreeLeaf<AstartePropertyData>({
+            interface: iface,
+            data: {
+              endpoint: `${endpoint}/${prop}`,
+              ...({
+                value: propValue,
+                type: getEndpointDataType(iface, `${endpoint}/${prop}`),
+              } as AstarteDataTuple),
+            },
+            endpoint: `${endpoint}/${prop}`,
+            relativeEndpoint: prop,
+            parentNode: this as AstarteDataTreeRoot<AstartePropertyData>,
+          });
+        }
+        if (
```

Comment:
here you check if the same interface is property or datastream for every leaf
you can check the interface value outside this method, and decide how to process all the leaves with a single map funciton

Reply:
I thought about that too. But extracting the logic would require a lot passing around of parameters, producing more complex code. I am also leveraging Typescript's generics and type inference; extracting the logic would require redefining and handling types, producing an even more verbose code.

Reply:
Yea, you SHOULD pass around parameters of informations you aready have

```
if (isPropObject(obj) {
  this.children = Object.entries(data).map(objToProvValue)
} else if (isDatastreamObject(obj)) {
  this.children = Object.entries(data).map(objToDatastreamValue)
} else ....
```